// ngx-captcha/ngx-dist/src/ngx-captcha.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("@angular/common");
var core_1 = require("@angular/core");
var invisible_recaptcha_component_1 = require("./recaptcha/invisible-recaptcha.component");
var recaptcha_2_component_1 = require("./recaptcha/recaptcha-2.component");
var recaptcha_config_1 = require("./recaptcha/recaptcha.config");
var NgxCaptchaModule = /** @class */ (function () {
    function NgxCaptchaModule() {
    }
    NgxCaptchaModule.forRoot = function (config) {
        return {
            ngModule: NgxCaptchaModule,
            providers: [
                {
                    provide: recaptcha_config_1.NgxCaptchaConfig,
                    useValue: config,
                }
            ]
        };
    };
    NgxCaptchaModule.decorators = [
        { type: core_1.NgModule, args: [{
                    imports: [
                        common_1.CommonModule
                    ],
                    declarations: [
                        recaptcha_2_component_1.ReCaptcha2Component,
                        invisible_recaptcha_component_1.InvisibleReCaptchaComponent
                    ],
                    exports: [
                        recaptcha_2_component_1.ReCaptcha2Component,
                        invisible_recaptcha_component_1.InvisibleReCaptchaComponent
                    ]
                },] },
    ];
    /** @nocollapse */
    NgxCaptchaModule.ctorParameters = function () { return []; };
    return NgxCaptchaModule;
}());
exports.NgxCaptchaModule = NgxCaptchaModule;
